mesonbuild/coredata.py

```diff
@@ -914,6 +922,11 @@ def update_project_options(self, options: 'MutableKeyedOptionDictType') -> None:
                     mlog.warning(f'Old value(s) of {key} are no longer valid, resetting to default ({value.value}).',
                                  fatal=False)
 
+        # Find any extranious keys for this project and remove them
+        for key in list(self.options.keys() - options.keys()):
+            if key.is_project() and key.subproject == subproject:
+                del self.options[key]
```

Comment:
Why do we care about making this a list? We can iterate either way...

Reply:
To make a copy, otherwise removing keys may mutate the list we’re iterating

Reply:
cef629c63924a48e24ebed51c7f7dcbfebaade96

> micro-optimize iteration of evaluated set
> 
> It is generally accepted practice to convert dict.keys() to a list before iterating over it and e.g. deleting values, as keys returns a live-action view. In this case, we use the difference of *two* dict keys, which returns a regular non-view set and doesn't need protecting.
> 
> Iteration order doesn't matter (the set already randomizes it anyway). Avoid the cost of converting to a list.